/**
 * @file photopage.cpp
 * @brief ç…§ç‰‡é¡µé¢ç»„ä»¶å®ç°
 */

#include "photopage.h"
#include "ui_photopage.h"
#include "flowlayout.h"

#include <QTreeWidgetItem>
#include <QPainter>
#include <QMouseEvent>
#include <QFileDialog>
#include <QMessageBox>
#include <QApplication>
#include <QVBoxLayout>
#include <QDebug>
#include <QTimer>
#include <QProgressDialog>
#include <QFile>
#include <QDir>

/* ============================================================================
 * PhotoThumbnail å®ç°
 * ============================================================================ */

PhotoThumbnail::PhotoThumbnail(const PhotoInfo &info, QWidget *parent)
    : QWidget(parent)
    , m_photoInfo(info)
    , m_selected(false)
    , m_hovered(false)
{
    setFixedSize(120, 120);
    setCursor(Qt::PointingHandCursor);
    
    // è®¾ç½®é»˜è®¤ç¼©ç•¥å›¾ï¼ˆå ä½ç¬¦ï¼‰
    m_thumbnail = QPixmap(100, 100);
    m_thumbnail.fill(Qt::lightGray);
}

void PhotoThumbnail::setSelected(bool selected)
{
    if (m_selected != selected) {
        m_selected = selected;
        update();
        emit selectionChanged(selected);
    }
}

void PhotoThumbnail::setThumbnail(const QPixmap &pixmap)
{
    if (!pixmap.isNull()) {
        m_thumbnail = pixmap.scaled(100, 100, Qt::KeepAspectRatio, Qt::SmoothTransformation);
        update();
    }
}

void PhotoThumbnail::paintEvent(QPaintEvent *event)
{
    Q_UNUSED(event)
    
    QPainter painter(this);
    painter.setRenderHint(QPainter::Antialiasing);
    
    // èƒŒæ™¯
    QRect bgRect = rect().adjusted(2, 2, -2, -2);
    if (m_selected) {
        painter.fillRect(bgRect, QColor("#e3f2fd"));
        painter.setPen(QPen(QColor("#1976d2"), 2));
        painter.drawRect(bgRect);
    } else if (m_hovered) {
        painter.fillRect(bgRect, QColor("#f5f5f5"));
        painter.setPen(QPen(QColor("#e0e0e0"), 1));
        painter.drawRect(bgRect);
    } else {
        painter.fillRect(bgRect, Qt::white);
        painter.setPen(QPen(QColor("#e8e8e8"), 1));
        painter.drawRect(bgRect);
    }
    
    // ç¼©ç•¥å›¾
    int imgX = (width() - m_thumbnail.width()) / 2;
    int imgY = (height() - m_thumbnail.height()) / 2 - 8;
    painter.drawPixmap(imgX, imgY, m_thumbnail);
    
    // æ–‡ä»¶å
    QRect textRect(4, height() - 24, width() - 8, 20);
    painter.setPen(Qt::black);
    QFont font = painter.font();
    font.setPointSize(8);
    painter.setFont(font);
    QString elidedText = painter.fontMetrics().elidedText(
        m_photoInfo.name, Qt::ElideMiddle, textRect.width());
    painter.drawText(textRect, Qt::AlignCenter, elidedText);
    
    // è§†é¢‘æ ‡è¯†
    if (m_photoInfo.isVideo) {
        QRect videoRect(width() - 24, 6, 18, 18);
        painter.fillRect(videoRect, QColor(0, 0, 0, 128));
        painter.setPen(Qt::white);
        painter.drawText(videoRect, Qt::AlignCenter, "â–¶");
    }
    
    // é€‰ä¸­æ ‡è®°ï¼ˆå¿ƒå½¢å›¾æ ‡ï¼‰
    if (m_selected || m_hovered) {
        QRect heartRect(8, height() - 32, 16, 16);
        painter.setPen(m_selected ? QColor("#e65100") : QColor("#999999"));
        painter.setFont(QFont("Arial", 12));
        painter.drawText(heartRect, Qt::AlignCenter, m_selected ? "â™¥" : "â™¡");
    }
}

void PhotoThumbnail::mousePressEvent(QMouseEvent *event)
{
    if (event->button() == Qt::LeftButton) {
        setSelected(!m_selected);
        emit clicked();
    }
    QWidget::mousePressEvent(event);
}

void PhotoThumbnail::mouseDoubleClickEvent(QMouseEvent *event)
{
    if (event->button() == Qt::LeftButton) {
        emit doubleClicked();
    }
    QWidget::mouseDoubleClickEvent(event);
}

void PhotoThumbnail::enterEvent(QEnterEvent *event)
{
    m_hovered = true;
    update();
    QWidget::enterEvent(event);
}

void PhotoThumbnail::leaveEvent(QEvent *event)
{
    m_hovered = false;
    update();
    QWidget::leaveEvent(event);
}

/* ============================================================================
 * PhotoPage å®ç°
 * ============================================================================ */

PhotoPage::PhotoPage(QWidget *parent)
    : QWidget(parent)
    , ui(new Ui::PhotoPage)
    , m_photoManager(nullptr)
    , m_flowLayout(nullptr)
    , m_libraryItem(nullptr)
    , m_albumsItem(nullptr)
{
    ui->setupUi(this);
    setupUI();
    setupAlbumTree();
}

PhotoPage::~PhotoPage()
{
    clearPhotoGrid();
    delete ui;
}

void PhotoPage::setupUI()
{
    // åˆ›å»ºæµå¼å¸ƒå±€ç”¨äºç…§ç‰‡ç½‘æ ¼
    // æ³¨æ„: UI æ–‡ä»¶ä¸­ photoGridContainer å·²ç»æœ‰ QVBoxLayout (photoGridLayout)
    // è¿™é‡Œå°† FlowLayout ä½œä¸ºå­å¸ƒå±€è¿½åŠ åˆ°è¯¥ QVBoxLayout ä¸­ï¼Œé¿å… setLayout å¤±è´¥å¯¼è‡´ FlowLayout æœªè¢«å®‰è£…
    {
        QLayout* containerLayout = ui->photoGridContainer->layout();
        if (auto vbox = qobject_cast<QVBoxLayout*>(containerLayout)) {
            m_flowLayout = new FlowLayout(nullptr, 8, 8, 8);
            vbox->addLayout(m_flowLayout);
            qDebug() << "[PhotoPage] setupUI: ä½¿ç”¨å·²æœ‰çš„ QVBoxLayout ä½œä¸ºå®¹å™¨"
                     << "vbox=" << vbox << "flow=" << m_flowLayout;
        } else {
            // å…œåº•: è‹¥æ— å¸ƒå±€åˆ™ç›´æ¥è®¾ç½® FlowLayout
            m_flowLayout = new FlowLayout(ui->photoGridContainer, 8, 8, 8);
            ui->photoGridContainer->setLayout(m_flowLayout);
            qDebug() << "[PhotoPage] setupUI: ç›´æ¥è®¾ç½® FlowLayout ä¸ºå®¹å™¨å¸ƒå±€"
                     << "flow=" << m_flowLayout;
        }
    }
    
    // è¿æ¥ä¿¡å·
    connect(ui->refreshButton, &QPushButton::clicked, this, &PhotoPage::onRefreshClicked);
    connect(ui->exportButton, &QPushButton::clicked, this, &PhotoPage::onExportClicked);
    connect(ui->albumTree, &QTreeWidget::currentItemChanged, this, &PhotoPage::onAlbumSelectionChanged);
}

void PhotoPage::setupAlbumTree()
{
    ui->albumTree->clear();
    
    // é‡ç½®é¡¹æŒ‡é’ˆ
    m_libraryItem = nullptr;
    m_albumsItem = nullptr;
    
    // å›¾åº“ - å›ºå®šé¡¹ï¼Œæ˜¾ç¤ºæ‰€æœ‰ç…§ç‰‡
    m_libraryItem = new QTreeWidgetItem(ui->albumTree);
    m_libraryItem->setText(0, "ğŸ“· å›¾åº“");
    m_libraryItem->setData(0, Qt::UserRole, "library");
    
    // æˆ‘çš„ç›¸ç°¿ - åŠ¨æ€åŠ è½½çš„ç›¸å†Œå®¹å™¨
    m_albumsItem = new QTreeWidgetItem(ui->albumTree);
    m_albumsItem->setText(0, "ğŸ“ æˆ‘çš„ç›¸ç°¿");
    m_albumsItem->setData(0, Qt::UserRole, "albums");
    m_albumsItem->setFlags(m_albumsItem->flags() | Qt::ItemIsAutoTristate);
    
    // é»˜è®¤é€‰ä¸­å›¾åº“
    ui->albumTree->setCurrentItem(m_libraryItem);
    
    // å±•å¼€æ‰€æœ‰é¡¹
    ui->albumTree->expandAll();
}

void PhotoPage::setPhotoManager(PhotoManager *manager)
{
    if (m_photoManager) {
        disconnect(m_photoManager, nullptr, this, nullptr);
    }
    
    m_photoManager = manager;
    
    if (m_photoManager) {
        connect(m_photoManager, &PhotoManager::scanProgress, this, &PhotoPage::onScanProgress);
        connect(m_photoManager, &PhotoManager::errorOccurred, this, &PhotoPage::onPhotoError);
    }
}

void PhotoPage::setCurrentDevice(const QString &udid)
{
    m_currentUdid = udid;
    ui->statusLabel->setText("è®¾å¤‡å·²è¿æ¥ï¼Œç‚¹å‡»åˆ·æ–°æŒ‰é’®åŠ è½½ç…§ç‰‡");
}

void PhotoPage::clearDevice()
{
    m_currentUdid.clear();
    m_currentAlbumPath.clear();
    clearPhotoGrid();
    updateStats(0, 0);
    ui->albumTitleLabel->setText("å…¨éƒ¨ç…§ç‰‡");
    ui->statusLabel->setText("è¯·å…ˆè¿æ¥è®¾å¤‡ä»¥æŸ¥çœ‹ç…§ç‰‡");
    
    // æ¸…é™¤ç›¸å†Œæ ‘ä¸­çš„åŠ¨æ€ç›¸å†Œ
    while (m_albumsItem->childCount() > 0) {
        delete m_albumsItem->takeChild(0);
    }
}

void PhotoPage::refreshPhotos()
{
    onRefreshClicked();
}

void PhotoPage::onRefreshClicked()
{
    if (m_currentUdid.isEmpty()) {
        QMessageBox::information(this, "æç¤º", "è¯·å…ˆè¿æ¥è®¾å¤‡");
        return;
    }
    
    if (!m_photoManager) {
        QMessageBox::warning(this, "é”™è¯¯", "ç…§ç‰‡ç®¡ç†å™¨æœªåˆå§‹åŒ–");
        return;
    }
    
    ui->statusLabel->setText("æ­£åœ¨åŠ è½½ç…§ç‰‡åˆ—è¡¨...");
    ui->refreshButton->setEnabled(false);
    QApplication::processEvents();
    
    // è¿æ¥åˆ°è®¾å¤‡
    if (!m_photoManager->isConnected()) {
        if (!m_photoManager->connectToDevice(m_currentUdid)) {
            ui->statusLabel->setText(QString("è¿æ¥å¤±è´¥: %1").arg(m_photoManager->lastError()));
            ui->refreshButton->setEnabled(true);
            return;
        }
    }
    
    // è·å–ç›¸å†Œåˆ—è¡¨
    QVector<AlbumInfo> albums = m_photoManager->getAlbums();
    updateAlbumTree(albums);
    
    // è·å–å½“å‰ç›¸å†Œçš„ç…§ç‰‡
    QVector<PhotoInfo> photos;
    if (m_currentAlbumPath.isEmpty()) {
        photos = m_photoManager->getAllPhotos();
    } else {
        photos = m_photoManager->getPhotos(m_currentAlbumPath);
    }
    
    if (photos.isEmpty() && !m_photoManager->lastError().isEmpty()) {
        ui->statusLabel->setText(QString("åŠ è½½å¤±è´¥: %1").arg(m_photoManager->lastError()));
    } else {
        displayPhotos(photos);
        ui->statusLabel->setText("å¦‚è¦åˆ é™¤æ­¤ç›®å½•å†…ç…§ç‰‡ï¼Œè¯·åˆ°è®¾å¤‡ä¸Šæ“ä½œã€‚");
    }
    
    ui->refreshButton->setEnabled(true);
}

void PhotoPage::onExportClicked()
{
    QVector<PhotoInfo> selected = getSelectedPhotos();
    
    if (selected.isEmpty()) {
        QMessageBox::information(this, "æç¤º", "è¯·å…ˆé€‰æ‹©è¦å¯¼å‡ºçš„ç…§ç‰‡");
        return;
    }
    
    QString dir = QFileDialog::getExistingDirectory(this, "é€‰æ‹©å¯¼å‡ºç›®å½•");
    if (dir.isEmpty()) {
        return;
    }

    if (!m_photoManager || !m_photoManager->isConnected()) {
        QMessageBox::warning(this, "é”™è¯¯", "è®¾å¤‡æœªè¿æ¥");
        return;
    }

    // åˆ›å»ºè¿›åº¦å¯¹è¯æ¡†
    QProgressDialog progress("æ­£åœ¨å¯¼å‡ºç…§ç‰‡...", "å–æ¶ˆ", 0, selected.size(), this);
    progress.setWindowModality(Qt::WindowModal);
    progress.setMinimumDuration(0); // ç«‹å³æ˜¾ç¤º
    progress.setValue(0);

    int successCount = 0;
    int failCount = 0;
    QString lastError;

    for (int i = 0; i < selected.size(); ++i) {
        if (progress.wasCanceled()) {
            break;
        }

        const PhotoInfo& photo = selected[i];
        progress.setLabelText(QString("æ­£åœ¨å¯¼å‡º (%1/%2): %3").arg(i + 1).arg(selected.size()).arg(photo.name));
        
        // è¯»å–ç…§ç‰‡æ•°æ®
        QByteArray data = m_photoManager->readPhotoData(photo.path);
        if (data.isEmpty()) {
            failCount++;
            lastError = m_photoManager->lastError();
            qDebug() << "[PhotoPage] å¯¼å‡ºå¤±è´¥(è¯»å–é”™è¯¯):" << photo.path << lastError;
        } else {
            // å†™å…¥æœ¬åœ°æ–‡ä»¶
            QString targetPath = QDir(dir).filePath(photo.name);
            // å¦‚æœæ–‡ä»¶å·²å­˜åœ¨ï¼Œè‡ªåŠ¨é‡å‘½å: name_1.jpg, name_2.jpg
            if (QFile::exists(targetPath)) {
                QFileInfo fi(targetPath);
                int counter = 1;
                while (QFile::exists(targetPath)) {
                    targetPath = QDir(dir).filePath(QString("%1_%2.%3")
                        .arg(fi.baseName())
                        .arg(counter++)
                        .arg(fi.suffix()));
                }
            }

            QFile file(targetPath);
            if (file.open(QIODevice::WriteOnly)) {
                if (file.write(data) == data.size()) {
                    successCount++;
                } else {
                    failCount++;
                    lastError = "å†™å…¥æ–‡ä»¶å¤±è´¥";
                    qDebug() << "[PhotoPage] å¯¼å‡ºå¤±è´¥(å†™å…¥é”™è¯¯):" << targetPath;
                }
                file.close();
            } else {
                failCount++;
                lastError = "æ— æ³•åˆ›å»ºç›®æ ‡æ–‡ä»¶";
                qDebug() << "[PhotoPage] å¯¼å‡ºå¤±è´¥(åˆ›å»ºæ–‡ä»¶é”™è¯¯):" << targetPath;
            }
        }

        progress.setValue(i + 1);
        QApplication::processEvents(); // ä¿æŒç•Œé¢å“åº”
    }

    progress.close();

    QString resultMsg = QString("å¯¼å‡ºå®Œæˆ\næˆåŠŸ: %1\nå¤±è´¥: %2").arg(successCount).arg(failCount);
    if (failCount > 0 && !lastError.isEmpty()) {
        resultMsg += QString("\n\næœ€åä¸€æ¬¡é”™è¯¯: %1").arg(lastError);
    }

    QMessageBox::information(this, "å¯¼å‡ºç»“æœ", resultMsg);
}

void PhotoPage::onAlbumSelectionChanged(QTreeWidgetItem *current, QTreeWidgetItem *previous)
{
    Q_UNUSED(previous)
    
    if (!current) {
        qDebug() << "[PhotoPage] onAlbumSelectionChanged: current is null";
        return;
    }
    
    QString albumType = current->data(0, Qt::UserRole).toString();
    QString albumPath = current->data(0, Qt::UserRole + 1).toString();
    
    qDebug() << "[PhotoPage] onAlbumSelectionChanged:"
             << "text=" << current->text(0)
             << "type=" << albumType
             << "path=" << albumPath;
    
    // æ›´æ–°æ ‡é¢˜
    ui->albumTitleLabel->setText(current->text(0));
    
    // æ›´æ–°å½“å‰ç›¸å†Œè·¯å¾„
    if (albumType == "library" || albumType == "albums") {
        m_currentAlbumPath.clear(); // æ˜¾ç¤ºå…¨éƒ¨
        qDebug() << "[PhotoPage] -> æ˜¾ç¤ºå…¨éƒ¨ç…§ç‰‡ (m_currentAlbumPath cleared)";
    } else if (albumType == "album") {
        m_currentAlbumPath = albumPath;
        qDebug() << "[PhotoPage] -> æ˜¾ç¤ºç›¸å†Œ:" << m_currentAlbumPath;
    } else {
        qDebug() << "[PhotoPage] -> æœªçŸ¥ç±»å‹ï¼Œä¸å¤„ç†";
        return;
    }
    
    // å¦‚æœå·²è¿æ¥è®¾å¤‡ï¼Œåªåˆ·æ–°ç…§ç‰‡ï¼ˆä¸é‡æ–°åŠ è½½ç›¸å†Œæ ‘ï¼‰
    if (!m_currentUdid.isEmpty() && m_photoManager && m_photoManager->isConnected()) {
        qDebug() << "[PhotoPage] -> è°ƒç”¨ loadPhotosForCurrentAlbum()";
        loadPhotosForCurrentAlbum();
    } else {
        qDebug() << "[PhotoPage] -> è·³è¿‡åŠ è½½: udid=" << m_currentUdid
                 << "manager=" << (m_photoManager ? "æœ‰" : "æ— ")
                 << "connected=" << (m_photoManager ? m_photoManager->isConnected() : false);
    }
}

void PhotoPage::loadPhotosForCurrentAlbum()
{
    qDebug() << "[PhotoPage] loadPhotosForCurrentAlbum: albumPath=" << m_currentAlbumPath;
    
    if (!m_photoManager || !m_photoManager->isConnected()) {
        qDebug() << "[PhotoPage] loadPhotosForCurrentAlbum: manageræ— æ•ˆæˆ–æœªè¿æ¥";
        return;
    }
    
    ui->statusLabel->setText("æ­£åœ¨åŠ è½½ç…§ç‰‡åˆ—è¡¨...");
    QApplication::processEvents();
    
    // è·å–å½“å‰ç›¸å†Œçš„ç…§ç‰‡
    QVector<PhotoInfo> photos;
    if (m_currentAlbumPath.isEmpty()) {
        qDebug() << "[PhotoPage] è°ƒç”¨ getAllPhotos()";
        photos = m_photoManager->getAllPhotos();
    } else {
        qDebug() << "[PhotoPage] è°ƒç”¨ getPhotos(" << m_currentAlbumPath << ")";
        photos = m_photoManager->getPhotos(m_currentAlbumPath);
    }
    
    qDebug() << "[PhotoPage] è·å–åˆ°ç…§ç‰‡æ•°é‡:" << photos.size()
             << "lastError:" << m_photoManager->lastError();
    
    if (photos.isEmpty() && !m_photoManager->lastError().isEmpty()) {
        ui->statusLabel->setText(QString("åŠ è½½å¤±è´¥: %1").arg(m_photoManager->lastError()));
    } else {
        displayPhotos(photos);
        ui->statusLabel->setText("å¦‚è¦åˆ é™¤æ­¤ç›®å½•å†…ç…§ç‰‡ï¼Œè¯·åˆ°è®¾å¤‡ä¸Šæ“ä½œã€‚");
    }
}

void PhotoPage::onScanProgress(int current, int total)
{
    ui->statusLabel->setText(QString("æ­£åœ¨æ‰«æ: %1 / %2").arg(current).arg(total));
    emit loadProgress(current, total);
    QApplication::processEvents();
}

void PhotoPage::onPhotoError(const QString &error)
{
    ui->statusLabel->setText(QString("é”™è¯¯: %1").arg(error));
    emit errorOccurred(error);
}

void PhotoPage::updateAlbumTree(const QVector<AlbumInfo> &albums)
{
    // ä¿å­˜å½“å‰é€‰ä¸­çš„ç›¸å†Œè·¯å¾„ï¼Œä»¥ä¾¿åˆ·æ–°åæ¢å¤é€‰ä¸­
    QString currentPath = m_currentAlbumPath;
    QTreeWidgetItem *itemToSelect = nullptr;

    // æš‚æ—¶å±è”½ä¿¡å·ï¼Œé˜²æ­¢åœ¨é‡å»ºæ ‘çš„è¿‡ç¨‹ä¸­è§¦å‘ä¸å¿…è¦çš„ selectionChanged
    // ä»è€Œå¯¼è‡´ m_currentAlbumPath è¢«é”™è¯¯é‡ç½®
    const bool wasBlocked = ui->albumTree->blockSignals(true);

    // æ¸…é™¤ç°æœ‰ç›¸å†Œ
    while (m_albumsItem->childCount() > 0) {
        delete m_albumsItem->takeChild(0);
    }
    
    // æ·»åŠ ç›¸å†Œ
    for (const AlbumInfo &album : albums) {
        QTreeWidgetItem *item = new QTreeWidgetItem(m_albumsItem);
        item->setText(0, album.name);
        item->setData(0, Qt::UserRole, "album");
        item->setData(0, Qt::UserRole + 1, album.path);
        
        // å¦‚æœæœ‰ç…§ç‰‡æ•°é‡ï¼Œæ˜¾ç¤ºåœ¨åç§°å
        if (album.photoCount > 0) {
            item->setText(0, QString("%1 (%2)").arg(album.name).arg(album.photoCount));
        }

        // æ£€æŸ¥æ˜¯å¦æ˜¯ä¹‹å‰é€‰ä¸­çš„ç›¸å†Œ
        if (!currentPath.isEmpty() && album.path == currentPath) {
            itemToSelect = item;
        }
    }
    
    // å±•å¼€ç›¸ç°¿
    m_albumsItem->setExpanded(true);

    // æ¢å¤é€‰ä¸­çŠ¶æ€
    if (itemToSelect) {
        ui->albumTree->setCurrentItem(itemToSelect);
    }

    // æ¢å¤ä¿¡å·
    ui->albumTree->blockSignals(wasBlocked);
}

void PhotoPage::displayPhotos(const QVector<PhotoInfo> &photos)
{
    qDebug() << "[PhotoPage] displayPhotos: æ”¶åˆ°" << photos.size() << "å¼ ç…§ç‰‡";
    
    clearPhotoGrid();
    
    qDebug() << "[PhotoPage] displayPhotos: flowLayout=" << m_flowLayout
             << "parent=" << (m_flowLayout ? m_flowLayout->parentWidget() : nullptr)
             << "container=" << ui->photoGridContainer
             << "container.layout=" << (ui->photoGridContainer ? ui->photoGridContainer->layout() : nullptr)
             << "container.visible=" << (ui->photoGridContainer ? ui->photoGridContainer->isVisible() : false)
             << "container.size=" << (ui->photoGridContainer ? ui->photoGridContainer->size() : QSize());
    
    int photoCount = 0;
    int videoCount = 0;
    
    for (const PhotoInfo &photo : photos) {
        PhotoThumbnail *thumbnail = new PhotoThumbnail(photo, ui->photoGridContainer);
        m_flowLayout->addWidget(thumbnail);
        thumbnail->show(); // ç¡®ä¿å¯è§
        m_thumbnails.append(thumbnail);
        
        if (photo.isVideo) {
            videoCount++;
        } else {
            photoCount++;
        }
        
        // TODO: å¼‚æ­¥åŠ è½½ç¼©ç•¥å›¾
        // ç›®å‰ä½¿ç”¨å ä½ç¬¦
    }
    
    // æ¿€æ´»å®¹å™¨å¸ƒå±€ï¼Œå¼ºåˆ¶æ‰§è¡Œå­å¸ƒå±€å‡ ä½•æ›´æ–°
    if (ui->photoGridContainer->layout()) {
        ui->photoGridContainer->layout()->activate();
    }
    ui->photoGridContainer->updateGeometry();
    ui->photoGridContainer->update();
    
    qDebug() << "[PhotoPage] displayPhotos: æ·»åŠ äº†" << m_thumbnails.size() << "ä¸ªç¼©ç•¥å›¾"
             << "layoutCount=" << m_flowLayout->count();
    
    // æ£€æŸ¥ç¬¬ä¸€ä¸ªç¼©ç•¥å›¾çŠ¶æ€
    if (!m_thumbnails.isEmpty()) {
        PhotoThumbnail *first = m_thumbnails.first();
        qDebug() << "[PhotoPage] ç¬¬ä¸€ä¸ªç¼©ç•¥å›¾: parent=" << first->parentWidget()
                 << "visible=" << first->isVisible()
                 << "geometry=" << first->geometry();
    }
    
    updateStats(photoCount, videoCount);

    // å¼€å§‹å¼‚æ­¥åŠ è½½ç¼©ç•¥å›¾
    startThumbnailLoading();
}

void PhotoPage::clearPhotoGrid()
{
    // æ¸…ç©ºå¾…åŠ è½½é˜Ÿåˆ—ï¼Œé˜²æ­¢è®¿é—®å·²åˆ é™¤çš„å¯¹è±¡
    m_pendingThumbnails.clear();
    m_isLoadingThumbnails = false;

    for (PhotoThumbnail *thumbnail : m_thumbnails) {
        m_flowLayout->removeWidget(thumbnail);
        thumbnail->deleteLater();
    }
    m_thumbnails.clear();
}

void PhotoPage::startThumbnailLoading()
{
    // åœæ­¢å½“å‰çš„åŠ è½½ï¼ˆå¦‚æœæœ‰ï¼‰
    m_pendingThumbnails.clear();
    m_isLoadingThumbnails = false;
    
    // å°†æ‰€æœ‰éœ€è¦åŠ è½½ç¼©ç•¥å›¾çš„é¡¹ç›®åŠ å…¥é˜Ÿåˆ—
    // æš‚æ—¶è·³è¿‡è§†é¢‘æ–‡ä»¶ï¼Œå› ä¸ºè¯»å–å¤§æ–‡ä»¶ä¼šé˜»å¡ï¼Œä¸” QImage æ— æ³•è§£ç è§†é¢‘
    for (PhotoThumbnail *thumbnail : m_thumbnails) {
        if (!thumbnail->photoInfo().isVideo) {
            m_pendingThumbnails.append(thumbnail);
        }
    }
    
    if (!m_pendingThumbnails.isEmpty()) {
        m_isLoadingThumbnails = true;
        // ä½¿ç”¨ 0ms å®šæ—¶å™¨åœ¨ä¸‹ä¸€æ¬¡äº‹ä»¶å¾ªç¯å¼€å§‹åŠ è½½ï¼Œé¿å…é˜»å¡å½“å‰æ“ä½œ
        QTimer::singleShot(0, this, &PhotoPage::loadNextThumbnail);
    }
}

void PhotoPage::loadNextThumbnail()
{
    if (m_pendingThumbnails.isEmpty()) {
        m_isLoadingThumbnails = false;
        qDebug() << "[PhotoPage] ç¼©ç•¥å›¾åŠ è½½å®Œæˆ";
        return;
    }

    // å–å‡ºä¸‹ä¸€ä¸ªè¦åŠ è½½çš„ç¼©ç•¥å›¾
    // æ³¨æ„ï¼šè¿™é‡Œéœ€è¦ä½¿ç”¨ QPointer æˆ–è€…ç¡®ä¿ clearPhotoGrid æ—¶æ¸…ç©ºäº†é˜Ÿåˆ—
    // æˆ‘ä»¬å·²ç»åœ¨ clearPhotoGrid ä¸­å¤„ç†äº†æ¸…ç©ºï¼Œæ‰€ä»¥è¿™é‡Œåº”è¯¥æ˜¯å®‰å…¨çš„
    PhotoThumbnail *thumbnail = m_pendingThumbnails.takeFirst();
    
    // åŒé‡æ£€æŸ¥ï¼šç¡®ä¿ thumbnail ä»ç„¶æœ‰æ•ˆï¼ˆè™½ç„¶æˆ‘ä»¬æ¸…ç©ºäº†é˜Ÿåˆ—ï¼Œä½†åœ¨æŸäº›æç«¯æƒ…å†µä¸‹ä¸ºäº†å®‰å…¨ï¼‰
    if (thumbnail && m_thumbnails.contains(thumbnail)) {
        QString path = thumbnail->photoInfo().path;
        QString ext = QFileInfo(path).suffix().toLower();
        
        // è¯»å–ç…§ç‰‡æ•°æ®
        // æ³¨æ„ï¼šè¿™æ˜¯åŒæ­¥è¯»å–ï¼Œå¯¹äºå¤§æ–‡ä»¶å¯èƒ½ä¼šæœ‰è½»å¾®å¡é¡¿
        QByteArray data = m_photoManager->readPhotoData(path);
        
        if (!data.isEmpty()) {
            QImage image;
            // å°è¯•ä»æ•°æ®åŠ è½½å›¾ç‰‡
            if (image.loadFromData(data)) {
                // ç¼©æ”¾å›¾ç‰‡ä»¥èŠ‚çœå†…å­˜
                QPixmap pixmap = QPixmap::fromImage(image);
                thumbnail->setThumbnail(pixmap);
            } else {
                qDebug() << "[PhotoPage] å›¾ç‰‡è§£ç å¤±è´¥:" << path
                         << "æ ¼å¼:" << ext
                         << "æ•°æ®å¤§å°:" << formatFileSize(data.size());
                
                if (ext == "heic" || ext == "heif") {
                    qDebug() << "[PhotoPage] æç¤º: Qt å¯èƒ½ç¼ºå°‘ HEIC/HEIF å›¾åƒæ ¼å¼æ’ä»¶";
                }
            }
        } else {
             qDebug() << "[PhotoPage] è¯»å–æ•°æ®å¤±è´¥(ç©º):" << path << "é”™è¯¯:" << m_photoManager->lastError();
        }
    }
    
    // ç»§ç»­åŠ è½½ä¸‹ä¸€ä¸ª
    // ä½¿ç”¨ 1ms å»¶è¿Ÿç»™ UI çº¿ç¨‹å–˜æ¯çš„æœºä¼šï¼Œä¿æŒç•Œé¢å“åº”
    if (m_isLoadingThumbnails) {
        QTimer::singleShot(1, this, &PhotoPage::loadNextThumbnail);
    }
}

void PhotoPage::updateStats(int photoCount, int videoCount)
{
    ui->photoCountLabel->setText(
        QString("%1 å¼ ç…§ç‰‡ï¼Œ%2 ä¸ªè§†é¢‘").arg(photoCount).arg(videoCount));
}

QVector<PhotoInfo> PhotoPage::getSelectedPhotos() const
{
    QVector<PhotoInfo> selected;
    for (PhotoThumbnail *thumbnail : m_thumbnails) {
        if (thumbnail->isSelected()) {
            selected.append(thumbnail->photoInfo());
        }
    }
    return selected;
}

QString PhotoPage::formatFileSize(qint64 size) const
{
    if (size < 1024) {
        return QString("%1 B").arg(size);
    } else if (size < 1024 * 1024) {
        return QString("%1 KB").arg(size / 1024.0, 0, 'f', 1);
    } else if (size < 1024 * 1024 * 1024) {
        return QString("%1 MB").arg(size / (1024.0 * 1024.0), 0, 'f', 1);
    } else {
        return QString("%1 GB").arg(size / (1024.0 * 1024.0 * 1024.0), 0, 'f', 2);
    }
}